vm.js

```diff
@@ -898,60 +900,154 @@ Object.subclass('Squeak.Image',
                     throw Error("Unexpected free block");
             }
             nWords--;  //length includes base header which we have already read
-            var oop = pos - 4 - headerSize, //0-rel byte oop of this object (base header)
+            let oop = pos - 4 - headerSize, //0-rel byte oop of this object (base header)
                 format = (header>>8) & 15,
                 hash = (header>>17) & 4095,
                 bits = readBits(nWords, format);
 
-            var object = new Squeak.Object();
-            object.initFromImage(oop, classInt, format, hash, bits);
-            if (classInt < 32) object.hash |= 0x10000000;    // see fixCompactOops()
-            if (prevObj) prevObj.nextObject = object;
+            let newOop = oldBaseAddr + oop;
+            if(!splObs && (newOop === specialObjectsOopInt)) {
+                splObs = bits;
+                floatClassOop = splObs[Squeak.splOb_ClassFloat];
+                compactClassesOop = splObs[Squeak.splOb_CompactClasses];
+            } else if(!compactClasses && (newOop === compactClassesOop)) {
+                compactClasses = bits;
+            }
+            let previousOop = this.previousOop;
+            this.previousOop = newOop;
+            Object.defineProperty(oopMap, newOop, {configurable: true, get: function(){
```

Comment:
We need to find a better way to write this. The read routine is already very complex, and this makes it even harder to understand.
Also, `newOop` is actually the old oop. New oops (`obj.oop`) are zero-based, old oops (in `obj.bits`) start at `oldBaseAddr`.